File: operators/compute/basic/layernorm.py
# Copyright (c) 2024 The Core Team
#
# Licensed under the Apache License, Version 2.0
# You may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an 'AS IS' BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import math
import torch
from typing import Tuple, Union, List

from . import init

_shape_t = Union[int, List[int], torch.Size]

def layernorm_init(
    normalized_shape: _shape_t, 
    eps: float = 1e-5, 
    elementwise_affine: bool = True,
    bias: bool = True, 
    device: torch.device = 'cuda:0', 
    dtype=torch.float32, 
    training=True,
):
    if len(normalized_shape) != 1:
        raise NotImplementedError("Currently layernorm only support normalized_dim=-1.")
    if elementwise_affine:
        is_bias = bias
        weight = torch.ones(normalized_shape, device=device, dtype=dtype)
        bias = torch.zeros(normalized_shape, device=device, dtype=dtype) if is_bias else None
        if training:
            weight.grad = torch.zeros_like(weight, device=weight.device, dtype=weight.dtype)
            bias.grad = torch.zeros_like(bias, device=bias.device, dtype=bias.dtype) if is_bias else None
    else:
        raise NotImplementedError("Currently layernorm only support elementwise_affine=True")
    init_info = (normalized_shape[0], eps, elementwise_affine)
    return weight, bias, init_info

def layernorm_forward(*args, backend="naive", **kwargs):
    if backend == "naive":
        return layernorm_forward_naive(*args, **kwargs)
    elif backend == "torchjit":
        return layernorm_forward_torchjit(*args, **kwargs)
    # elif backend == "triton":
    #     return layernorm_forward_triton(*args, **kwargs)

def layernorm_backward(*args, backend="naive", **kwargs):
    if backend == "naive":
        return layernorm_backward_naive(*args, **kwargs)
    elif backend == "torchjit":
        return layernorm_backward_torchjit(*args, **kwargs)
    # elif backend == "triton":
    #     return layernorm_backward_triton(*args, **kwargs)



def layernorm_forward_naive(input: torch.Tensor, weight: torch.Tensor, bias: torch.Tensor, init_info: tuple):
    normalized_shape, eps, elementwise_affine = init_info
    last_dims = -1
    mean = input.mean(dim=last_dims, keepdim=True)
    var = input.var(dim=last_dims, keepdim=True, unbiased=False)
    output_buffer = (input - mean) / torch.sqrt(var + eps)
    if elementwise_affine:
        output = output_buffer * weight + bias
    save_for_backward = (output, output_buffer, normalized_shape, mean, var, eps, elementwise_affine)
    return output, save_for_backward

def layernorm_backward_naive(grad_output: torch.Tensor, weight: torch.Tensor, bias: torch.Tensor, save_for_backward: tuple):
    output, output_buffer, normalized_shape, mean, var, eps, elementwise_affine = save_for_backward
    # Compute the normalization dimensions based on normalized_shape
    norm_dim = -1  # This computes it as the last dimensions
    non_norm_dims = tuple(range(len(grad_output.shape)-1))  # This computes it as the last dimensions

    if elementwise_affine:
        # Compute gradients for weight and bias
        grad_weight = (grad_output * output_buffer).sum(dim=non_norm_dims, keepdim=True)
        weight.grad += grad_weight.squeeze()
        grad_bias = grad_output.sum(dim=non_norm_dims, keepdim=True)
        bias.grad += grad_bias.squeeze()
        # Prepare grad_output for input gradient calculation
        grad_output = grad_output * weight

    # Gradient with respect to the input
    std_inv = 1 / torch.sqrt(var + eps)
    grad_input = grad_output * std_inv

    # Additional corrections for the mean and variance contributions
    grad_mean = (-std_inv * grad_output).sum(dim=norm_dim, keepdim=True)
    grad_var = (-0.5 * std_inv.pow(3) * (output - mean) * grad_output).sum(dim=norm_dim, keepdim=True)
    grad_input += (grad_var * 2 / torch.tensor(normalized_shape).prod() * (output - mean)) + \
                  (grad_mean / torch.tensor(normalized_shape).prod())

    return grad_input, weight if elementwise_affine else None, bias if elementwise_affine else None



def layernorm_forward_torchjit(input: torch.Tensor, weight: torch.Tensor, bias: torch.Tensor, init_info: tuple):
    normalized_shape, eps, elementwise_affine = init_info
    @torch.jit.script
    def compute(input: torch.Tensor, weight: torch.Tensor, bias: torch.Tensor, elementwise_affine: bool, eps: float):
        mean = input.mean(dim=-1, keepdim=True)
        var = input.var(dim=-1, keepdim=True, unbiased=False)
        output = (input - mean) / torch.sqrt(var + eps)
        output = output * weight + bias
        return output
    save_for_backward = (input, normalized_shape, eps, elementwise_affine)
    return compute(input, weight, bias, elementwise_affine, eps), save_for_backward

def layernorm_backward_torchjit(grad_output: torch.Tensor, weight: torch.Tensor, bias: torch.Tensor, save_for_backward: tuple):
    input, normalized_shape, eps, elementwise_affine = save_for_backward
    # Compute the normalization dimensions based on normalized_shape
    @torch.jit.script
    def compute(grad_output: torch.Tensor, input: torch.Tensor, weight: torch.Tensor, bias: torch.Tensor, normalized_shape: int, eps: float, elementwise_affine: bool):
        norm_dim = -1  # This computes it as the last dimensions
        # non_norm_dims = range(len(grad_output.shape)-1)  # This computes it as the last dimensions
        non_norm_dims = (i for i in range(len(grad_output.shape)-1))  # This computes it as the last dimensions
        mean = input.mean(dim=norm_dim, keepdim=True)
        var = input.var(dim=norm_dim, keepdim=True, unbiased=False)
        output_buffer = (input - mean) / torch.sqrt(var + eps)
        output = output_buffer * weight + bias
        grad_weight = (grad_output * output_buffer).sum(dim=non_norm_dims, keepdim=True)
        # weight.grad += grad_weight.squeeze()
        grad_bias = grad_output.sum(dim=non_norm_dims, keepdim=True)
        # bias.grad += grad_bias.squeeze()
        # Prepare grad_output for input gradient calculation
        grad_output = grad_output * weight
        # Gradient with respect to the input
        std_inv = 1 / torch.sqrt(var + eps)
        grad_input = grad_output * std_inv
        # Additional corrections for the mean and variance contributions
        grad_mean = (-std_inv * grad_output).sum(dim=norm_dim, keepdim=True)
        grad_var = (-0.5 * std_inv.pow(3) * (output - mean) * grad_output).sum(dim=norm_dim, keepdim=True)
        grad_input += (grad_var * 2 / torch.tensor(normalized_shape).prod() * (output - mean)) + \
                    (grad_mean / torch.tensor(normalized_shape).prod())
        return grad_input, grad_weight.squeeze(), grad_bias.squeeze()
    grad_input, weight.grad, bias.grad = compute(grad_output, input, weight, bias, normalized_shape, eps, elementwise_affine)
    return grad_input, weight, bias

File: src/compute/layer_norm.cpp
/*
 * Copyright (c) 2024 The Core team
 *
 * Licensed under the Apache License, Version 2.0;
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an 'AS IS' BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

// Header Order Protection
// ReSharper disable once CppUnusedIncludeDirective
#include <c10/util/Exception.h>
// Header Order Protection

#include <ATen/autocast_mode.h>
#include <ATen/ops/empty.h>
#include <ATen/ops/layer_norm.h>
#include <ATen/ops/native_layer_norm.h>
#include <ATen/ops/native_layer_norm_backward.h>
#include <ATen/ops/ones.h>
#include <ATen/ops/zeros.h>

#include "compute/layer_norm.h"
#include "logger.h"
#include "tensor_impl.h"
#include "threading/scheduler_impl.h"
#include "threading/task_impl.h"

namespace cs::compute {
LayerNorm::State::State(const Forward& forward, const Backward& backward,
                        const Args& args)
    : forward{forward}, backward{backward}, args{args} {}

OrderedDict<std::string, Tensor> LayerNorm::State::parameters() const {
  OrderedDict<std::string, Tensor> dict;
  dict.insert("weight", forward.weight);
  if (args.bias) {
    dict.insert("bias", forward.bias);
  }
  return dict;
}

OrderedDict<std::string, Tensor> LayerNorm::State::gradients() const {
  OrderedDict<std::string, Tensor> dict;
  dict.insert("weight", forward.grad_weight);
  if (args.bias) {
    dict.insert("bias", forward.grad_bias);
  }
  return dict;
}

OrderedDict<std::string, module::State::Increment>
LayerNorm::State::increments() const {
  OrderedDict<std::string, Increment> dict;
  dict.insert("weight", {forward.weight, forward.grad_weight});
  if (args.bias) {
    dict.insert("bias", {forward.bias, forward.grad_bias});
  }
  return dict;
}

void LayerNorm::State::zero_grad() {
  forward.grad_weight = {};
  if (args.bias) {
    forward.grad_bias = {};
  }
}

std::shared_ptr<LayerNorm::State> LayerNorm::init(const Scheduler& scheduler,
                                                  const Options& options) {
  CS_ASSERT_TRUE(options.elementwise_affine() == true,
                 "elementwise_affine must be enabled now");
  at::TensorOptions tensorOptions{};
  if (options.device().has_value()) {
    tensorOptions = tensorOptions.device(options.device());
  }
  if (options.dtype().has_value()) {
    tensorOptions = tensorOptions.dtype(options.dtype());
  }
  Tensor weight;
  if (options.bias()) {
    struct Impl : Task::Impl {
      const Options options;
      const TensorOptions tensorOptions;

      explicit Impl(std::vector<Tensor> output /* weight, bias */,
                    const Options& options, const TensorOptions tensorOptions)
          : Task::Impl{std::move(output), {}, kMain, kCompute},
            options{options},
            tensorOptions{tensorOptions} {}
      void operator()() const override {
        output()[0].impl()->tensor() =
            at::ones(options.normalized_shape(), tensorOptions);
        output()[1].impl()->tensor() =
            at::zeros(options.normalized_shape(), tensorOptions);
      }
      [[nodiscard]] const char* name() const override {
        return "cs::compute::LayerNorm::init";
      }
    };

    Tensor bias;
    auto task = Task{
        std::make_shared<Impl>(Impl{{weight, bias}, options, tensorOptions})};
    scheduler.impl()->submit(std::move(task));
    return std::make_shared<State>(
        State::Forward{std::move(weight), std::move(bias)}, State::Backward{},
        State::Args{options.normalized_shape(), options.eps(),
                    options.elementwise_affine(), options.bias()});
  } else {
    struct Impl : Task::Impl {
      const Options options;
      const TensorOptions tensorOptions;

      explicit Impl(std::vector<Tensor> output /* weight */,
                    const Options& options, const TensorOptions tensorOptions)
          : Task::Impl{std::move(output), {}, kMain, kCompute},
            options{options},
            tensorOptions{tensorOptions} {}
      void operator()() const override {
        output()[0].impl()->tensor() =
            at::ones(options.normalized_shape(), tensorOptions);
      }
      [[nodiscard]] const char* name() const override {
        return "cs::compute::LayerNorm::init";
      }
    };

    auto task =
        Task{std::make_shared<Impl>(Impl{{weight}, options, tensorOptions})};
    scheduler.impl()->submit(std::move(task));
    return std::make_shared<State>(
        State::Forward{std::move(weight)}, State::Backward{},
        State::Args{options.normalized_shape(), options.eps(),
                    options.elementwise_affine(), options.bias()});
  }
}

Tensor LayerNorm::forward(const Scheduler& scheduler,
                          const std::shared_ptr<State>& state,
                          const ReadOnlyTensor& input) {
  Tensor mean;
  Tensor rstd;
  Task task{nullptr};
  Tensor output;
  if (state->args.bias) {
    struct Impl : Task::Impl {
      State::Args args;

      explicit Impl(std::vector<Tensor> output /* output, mean, rstd */,
                    std::vector<ReadOnlyTensor> input /* input, weight, bias */,
                    const State::Args& args)
          : Task::Impl{std::move(output), std::move(input), kMain, kCompute},
            args{args} {}
      void operator()() const override {
        std::make_tuple(std::ref(output()[0].impl()->tensor()),
                        std::ref(output()[1].impl()->tensor()),
                        std::ref(output()[2].impl()->tensor())) =
            at::native_layer_norm(input()[0].impl()->tensor(),
                                  args.normalized_shape,
                                  input()[1].impl()->tensor(),
                                  input()[2].impl()->tensor(), args.eps);
        if (at::autocast::is_enabled()) {
          const auto dtype = at::autocast::get_autocast_gpu_dtype();
          input()[0].impl()->auto_cast().enable = true;
          input()[1].impl()->auto_cast().enable = true;
          output()[0].impl()->auto_cast().enable = true;
          output()[1].impl()->auto_cast().enable = true;
          output()[2].impl()->auto_cast().enable = true;
        }
      }
      [[nodiscard]] const char* name() const override {
        return "cs::compute::LayerNorm::forward";
      }
    };

    task = Task{std::make_shared<Impl>(
        Impl{{output, mean, rstd},
             {input, state->forward.weight, state->forward.bias},
             state->args})};
  } else {
    struct Impl : Task::Impl {
      State::Args args;

      explicit Impl(std::vector<Tensor> output /* output, mean, rstd */,
                    std::vector<ReadOnlyTensor> input /* input, weight */,
                    const State::Args& args)
          : Task::Impl{std::move(output), std::move(input), kMain, kCompute},
            args{args} {}
      void operator()() const override {
        std::make_tuple(std::ref(output()[0].impl()->tensor()),
                        std::ref(output()[1].impl()->tensor()),
                        std::ref(output()[2].impl()->tensor())) =
            at::native_layer_norm(input()[0].impl()->tensor(),
                                  args.normalized_shape,
                                  input()[1].impl()->tensor(), {}, args.eps);
        if (at::autocast::is_enabled()) {
          const auto dtype = at::autocast::get_autocast_gpu_dtype();
          input()[0].impl()->auto_cast().enable = true;
          input()[1].impl()->auto_cast().enable = true;
          output()[0].impl()->auto_cast().enable = true;
          output()[1].impl()->auto_cast().enable = true;
          output()[2].impl()->auto_cast().enable = true;
        }
      }
      [[nodiscard]] const char* name() const override {
        return "cs::compute::LayerNorm::forward";
      }
    };

    task = Task{std::make_shared<Impl>(Impl{
        {output, mean, rstd}, {input, state->forward.weight}, state->args})};
  }
  state->backward.input = input;
  state->backward.mean = mean;
  state->backward.rstd = rstd;
  scheduler.impl()->submit(std::move(task));
  return output;
}

Tensor LayerNorm::backward(const Scheduler& scheduler,
                           const std::shared_ptr<State>& state,
                           const ReadOnlyTensor& grad_output) {
  backward_parameter(scheduler, state, grad_output);
  return backward_input(scheduler, state, grad_output);
}

Tensor LayerNorm::backward_input(const Scheduler& scheduler,
                                 const std::shared_ptr<State>& state,
                                 const ReadOnlyTensor& grad_output) {
  auto weight = static_cast<const ReadOnlyTensor&>(state->forward.weight);
  Tensor grad_input;
  Task task{nullptr};
  struct Impl : Task::Impl {
    const State::Args args;

    explicit Impl(std::vector<Tensor> output /* grad_input */,
                  std::vector<ReadOnlyTensor>
                      input /* grad_output, input, mean, rstd, weight */,
                  const State::Args& args)
        : Task::Impl{std::move(output), std::move(input), kMain, kCompute},
          args{args} {}
    void operator()() const override {
      at::Tensor grad_output = input()[0].impl()->tensor();
      at::Tensor x = input()[1].impl()->tensor();
      if (input()[1].impl()->auto_cast().enable) {
        if (x.dtype() != c10::kFloat) {
          x = at::autocast::cached_cast(c10::kFloat, x);
        }
        if (grad_output.dtype() != c10::kFloat) {
          intermediate().resize(2);
          intermediate().push_back(grad_output);
          grad_output = grad_output.to(c10::kFloat);
        }
      }
      auto [dx, dw, db] = at::native_layer_norm_backward(
          grad_output, x, args.normalized_shape, input()[2].impl()->tensor(),
          input()[3].impl()->tensor(), input()[4].impl()->tensor(), {},
          {true, false, false});
      output()[0].impl()->tensor() = dx;
    }
    [[nodiscard]] const char* name() const override {
      return "cs::compute::LayerNorm::backward_input";
    }
  };

  task = Task{std::make_shared<Impl>(
      Impl{{grad_input},
           {grad_output, state->backward.input, state->backward.mean,
            state->backward.rstd, state->forward.weight},
           state->args})};

  state->backward.input.reset();
  state->backward.mean.reset();
  state->backward.rstd.reset();
  scheduler.impl()->submit(std::move(task));
  return grad_input;
}

void LayerNorm::backward_parameter(const Scheduler& scheduler,
                                   const std::shared_ptr<State>& state,
                                   const ReadOnlyTensor& grad_output) {
  auto weight = static_cast<const ReadOnlyTensor&>(state->forward.weight);
  Tensor grad_input;
  Task task{nullptr};
  if (state->args.bias) {
    struct Impl : Task::Impl {
      const State::Args args;

      explicit Impl(
          std::vector<Tensor> output /* grad_weight, grad_bias */,
          std::vector<ReadOnlyTensor>
              input /* grad_output, input, mean, rstd, weight, bias */,
          const State::Args& args)
          : Task::Impl{std::move(output), std::move(input), kAssist, kCompute},
            args{args} {}
      void operator()() const override {
        at::Tensor grad_output = input()[0].impl()->tensor();
        at::Tensor x = input()[1].impl()->tensor();
        if (input()[1].impl()->auto_cast().enable) {
          if (x.dtype() != c10::kFloat) {
            x = at::autocast::cached_cast(c10::kFloat, x);
          }
          if (grad_output.dtype() != c10::kFloat) {
            intermediate().resize(3);
            intermediate().push_back(grad_output);
            grad_output = grad_output.to(c10::kFloat);
          } else {
            intermediate().resize(2);
          }
        } else {
          intermediate().resize(2);
        }
        auto [dx, dw, db] = at::native_layer_norm_backward(
            grad_output, x, args.normalized_shape, input()[2].impl()->tensor(),
            input()[3].impl()->tensor(), input()[4].impl()->tensor(),
            input()[5].impl()->tensor(), {false, true, true});
        if (output()[1].impl()->tensor().defined()) {
          output()[1].impl()->tensor() += dw;
        } else {
          output()[1].impl()->tensor() = dw;
        }
        if (output()[2].impl()->tensor().defined()) {
          output()[2].impl()->tensor() += db;
        } else {
          output()[2].impl()->tensor() = db;
        }
        intermediate().push_back(dw);
        intermediate().push_back(db);
      }
      [[nodiscard]] const char* name() const override {
        return "cs::compute::LayerNorm::backward_parameter";
      }
    };

    task = Task{std::make_shared<Impl>(
        Impl{{grad_input, state->forward.grad_weight, state->forward.grad_bias},
             {grad_output, state->backward.input, state->backward.mean,
              state->backward.rstd, state->forward.weight, state->forward.bias},
             state->args})};
  } else {
    struct Impl : Task::Impl {
      const State::Args args;

      explicit Impl(std::vector<Tensor> output /* grad_weight */,
                    std::vector<ReadOnlyTensor>
                        input /* grad_output, input, mean, rstd, weight */,
                    const State::Args& args)
          : Task::Impl{std::move(output), std::move(input), kAssist, kCompute},
            args{args} {}
      void operator()() const override {
        at::Tensor grad_output = input()[0].impl()->tensor();
        at::Tensor x = input()[1].impl()->tensor();
        if (input()[1].impl()->auto_cast().enable) {
          if (x.dtype() != c10::kFloat) {
            x = at::autocast::cached_cast(c10::kFloat, x);
          }
          if (grad_output.dtype() != c10::kFloat) {
            intermediate().resize(2);
            intermediate().push_back(grad_output);
            grad_output = grad_output.to(c10::kFloat);
          } else {
            intermediate().resize(1);
          }
        } else {
          intermediate().resize(1);
        }
        auto [dx, dw, db] = at::native_layer_norm_backward(
            grad_output, x, args.normalized_shape, input()[2].impl()->tensor(),
            input()[3].impl()->tensor(), input()[4].impl()->tensor(), {},
            {false, true, false});
        if (output()[1].impl()->tensor().defined()) {
          output()[1].impl()->tensor() += dw;
        } else {
          output()[1].impl()->tensor() = dw;
        }
        intermediate().push_back(dw);
      }
      [[nodiscard]] const char* name() const override {
        return "cs::compute::LayerNorm::backward_parameter";
      }
    };

    task = Task{std::make_shared<Impl>(
        Impl{{grad_input, state->forward.grad_weight},
             {grad_output, state->backward.input, state->backward.mean,
              state->backward.rstd, state->forward.weight},
             state->args})};
  }

  scheduler.impl()->submit(std::move(task));
}

}  // namespace cs::compute
